File: Android/GameControllerSimulator/socket-common-library/src/main/java/cn/maoyanluo/socket_common_library/SocketServerCallback.kt
package cn.maoyanluo.socket_common_library

import java.io.Closeable

interface SocketServerCallback<TSocket: Closeable> {

    fun onStartServerSuccess()
    fun onStartServerFailed(e: Exception)
    fun onStopServer()
    fun onForeverLoopException(e: Exception)
    fun createNewClientCallback(): ClientCallback<TSocket>
    fun onNewClientConnect(client: SocketServer.Client<TSocket>)
    fun onNewClientException(e: Exception)

    interface ClientCallback<TSocket: Closeable> {
        fun onSendDataException(client: SocketServer.Client<TSocket>, e: Exception, id: Int = -1)
        fun onDisconnect(client: SocketServer.Client<TSocket>)
        fun onDataReady(client: SocketServer.Client<TSocket>, data: ByteArray)
        fun onDataRevException(client: SocketServer.Client<TSocket>, e: Exception)
    }

}

File: Windows/GameControllerSimulator/SocketCommonLibrary/SocketServerCallback.cs
using System;

namespace SocketCommonLibrary
{
    public interface SocketServerCallback<TServerSocket, TSocket> where TServerSocket: class, IDisposable where TSocket : class, IDisposable
    {

        void OnStartServerSuccess();
        void OnStartServerFailed(Exception ex);
        void OnStopServer();
        void OnStopServerException(Exception ex);

        void OnForeverLoopException(Exception ex);

        SocketClientCallback<TSocket> CreateNewClientCallback();
        void OnClientConnect(SocketClient<TSocket> client);
        void OnNewClientException(Exception ex);

        void OnTaskException(Exception ex);

    }

    public interface SocketClientCallback<TSocket> where TSocket : class, IDisposable
    {
        void OnSendDataException(SocketClient<TSocket> client, Exception ex, int id = -1);
        void OnDisconnect(SocketClient<TSocket> client);
        void OnDataReady(SocketClient<TSocket> client, byte[] data);
        void OnDataRevException(SocketClient<TSocket> client, Exception ex);
        void OnTaskException(Exception ex);
    }

}
